iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java

```diff
@@ -0,0 +1,1251 @@
+                      COMMON_CONFIG.getPipeConsensusReceiverMaxWaitingTimeForEventsInMs()
+                          * COMMON_CONFIG.getPipeConsensusEventBufferSize(),
+                      TimeUnit.MILLISECONDS);
+
+              // If the buffer is not full after waiting timeout, we suppose that the sender will
+              // not send any more events at this time, that is, the sender has sent all events. At
+              // this point we apply the event at reqBuffer's peek
+              if (timeout
+                  && reqBuffer.size() < COMMON_CONFIG.getPipeConsensusEventBufferSize()
+                  && reqBuffer.first() != null
+                  && reqBuffer.first().equals(wrappedReq)) {
+                TPipeConsensusTransferResp resp = loadEvent(req);
+
+                if (resp != null
+                    && resp.getStatus().getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+                  onSuccess(wrappedReq.getCommitIndex());
+                  // signal all other reqs that may wait for this event
+                  condition.signalAll();
+                }
+                return resp;
+              }
+            } catch (InterruptedException e) {
+              LOGGER.warn(
+                  "PipeConsensus-ConsensusGroup-{}: current waiting is interrupted. onSyncedCommitIndex: {}. Exception: ",
+                  consensusGroupId,
+                  wrappedReq.getCommitIndex(),
+                  e);
+              Thread.currentThread().interrupt();
+              // Avoid infinite loop when RPC thread is killed by OS
+              return new TPipeConsensusTransferResp(
+                  RpcUtils.getStatus(
+                      TSStatusCode.SHUT_DOWN_ERROR,
+                      "RPC processor is interrupted by shutdown hook when wait on condition!"));
+            }
+          }
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    /**
+     * Reset all data to initial status and set connectorRebootTimes properly. This method is called
+     * when receiver identifies connector has rebooted.
+     */
+    private void resetWithNewestRebootTime(int connectorRebootTimes) {
+      LOGGER.info(
+          "PipeConsensus-ConsensusGroup-{}: receiver detected an newer rebootTimes, which indicates the leader has rebooted. receiver will reset all its data.",
+          consensusGroupId);
+      this.reqBuffer.clear();
+      this.onSyncedCommitIndex = -1;
+      // sync the follower's connectorRebootTimes with connector's actual rebootTimes
+      this.connectorRebootTimes = connectorRebootTimes;
+    }
+  }
+
+  /**
+   * Wrapped TPipeConsensusTransferReq for RequestExecutor.reqBuffer in order to save memory
+   * allocation. We don’t really need to hold a reference to TPipeConsensusTransferReq here, because
+   * we only need the commitId information of TPipeConsensusTransferReq in the
+   * RequestExecutor.reqBuffer.
+   */
+  private static class WrappedRequest {
```

Comment:
Better rename it to "RequestOrderId" or something related to "order"... "WrappedRequest" looks as if there's a request inside.